action/protocol/execution/evm/evmstatedbadapter_test.go

```diff
@@ -453,3 +454,69 @@ func TestPreimage(t *testing.T) {
 	require.NoError(err)
 	require.Equal([]byte("hen"), []byte(k))
 }
+
+func TestSortMap(t *testing.T) {
+	uniqueSlice := func(slice []string) []string {
+		// create a map with all the values as key
+		uniqMap := make(map[string]struct{})
+		for _, v := range slice {
+			uniqMap[v] = struct{}{}
+		}
+
+		// turn the map keys into a slice
+		uniqSlice := make([]string, 0, len(uniqMap))
+		for v := range uniqMap {
+			uniqSlice = append(uniqSlice, v)
+		}
+		return uniqSlice
+	}
+	require := require.New(t)
+
+	ctrl := gomock.NewController(t)
+	sm, err := initMockStateManager(ctrl)
+	require.NoError(err)
+	t.Run("before fix sort map", func(t *testing.T) {
+		stateDB := NewStateDBAdapter(sm, 1, true, false, hash.ZeroHash256)
+		size := 10
+
+		for i := 0; i < size; i++ {
+			addr := common.HexToAddress(identityset.Address(i).Hex())
+			stateDB.SetCode(addr, []byte("0123456789"))
+			stateDB.SetState(addr, k1, k2)
+		}
+		sn := stateDB.Snapshot()
+		caches := []string{}
+		for i := 0; i < size; i++ {
+			stateDB.RevertToSnapshot(sn)
+			s := ""
+			for _, c := range stateDB.cachedContract {
+				s += string(c.SelfState().Root[:])
+			}
+			caches = append(caches, s)
+		}
+		require.Greater(len(uniqueSlice(caches)), 1)
+	})
+
+	t.Run("after fix sort map", func(t *testing.T) {
+		stateDB := NewStateDBAdapter(sm, 1, true, false, hash.ZeroHash256, SortCachedContractsOption())
+		size := 10
+
+		for i := 0; i < size; i++ {
+			addr := common.HexToAddress(identityset.Address(i).Hex())
+			stateDB.SetCode(addr, []byte("0123456789"))
+			stateDB.SetState(addr, k1, k2)
+		}
+		sn := stateDB.Snapshot()
+		caches := []string{}
+		for i := 0; i < size; i++ {
+			stateDB.RevertToSnapshot(sn)
+			s := ""
+			for _, addr := range stateDB.cachedContractAddrs() {
+				c := stateDB.cachedContract[addr]
```

Comment:
why we calculate each contract's root? is there an entire root to check?

from your test result, it seems correct to do this. maybe confirm with Zhi as well